File: algorithms/06_javascript/geometry/point_in_polygon.js
// Ray-casting point-in-polygon test.
function inside(polygon, px, py) {
  let result = false;
  const n = polygon.length;
  let j = n - 1;
  for (let i = 0; i < n; i += 1) {
    const [xi, yi] = polygon[i];
    const [xj, yj] = polygon[j];
    if ((yi > py) !== (yj > py) && px < ((xj - xi) * (py - yi)) / (yj - yi) + xi) {
      result = !result;
    }
    j = i;
  }
  return result;
}

const square = [[0, 0], [4, 0], [4, 4], [0, 4]];
if (!inside(square, 2, 2) || inside(square, 5, 5)) throw new Error("ray casting failed");
console.log("point in polygon ok");

File: algorithms/11_golang/geometry/point_in_polygon.go
// Ray-casting point-in-polygon test.
package main

import "fmt"

func inside(poly [][2]float64, px, py float64) bool {
	in := false
	n := len(poly)
	j := n - 1
	for i := 0; i < n; i++ {
		xi, yi := poly[i][0], poly[i][1]
		xj, yj := poly[j][0], poly[j][1]
		if (yi > py) != (yj > py) && px < (xj-xi)*(py-yi)/(yj-yi)+xi {
			in = !in
		}
		j = i
	}
	return in
}

func main() {
	square := [][2]float64{{0, 0}, {4, 0}, {4, 4}, {0, 4}}
	if !inside(square, 2, 2) || inside(square, 5, 5) {
		panic("ray casting failed")
	}
	fmt.Println("point in polygon ok")
}
